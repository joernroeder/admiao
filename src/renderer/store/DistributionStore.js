import React, { createContext, useContext, useReducer, useEffect } from 'react'

const localStorageKey = 'DistributionState'

const initialState = {
	distributionType: undefined,
	distributionIdentifier: undefined,
}

const DistributionTypes = {
	LOCAL_PRINT: 'LOCAL_PRINT',
	REMOTE_PRINT: 'REMOTE_PRINT',
	DIGITAL: 'DIGITAL',
	FILE_SYSTEM: 'FILE_SYSTEM',
}

function reducer(state, action) {
	// sets Distribution Type
	if (Object.keys(DistributionTypes).includes(action.type)) {
		return {
			...state,
			distributionType: DistributionTypes[action.type],
			distributionIdentifier: action.payload || undefined,
		}
	}

	switch (action.type) {
		case 'RESET': {
			localStorage.removeItem(localStorageKey)

			return {
				...initialState,
			}
		}

		case 'SET_IDENTIFIER': {
			return {
				...state,
				distributionIdentifier: action.payload,
			}
		}

		default:
			return state
	}
}

const DistributionStateContext = createContext(undefined)
const DistributionDispatchContext = createContext(undefined)

function DistributionProvider({ children }) {
	const [state, dispatch] = useReducer(
		reducer,
		initialState,
		initialState => {
			const previousState = localStorage.getItem(localStorageKey)

			return previousState ? JSON.parse(previousState) : initialState
		}
	)

	// store distributionType in local storage.
	// This stores the distribution type to improve the backup process.
	useEffect(() => {
		if (state.isInitialState) {
			return
		}

		// filter out empty and initial states
		if (!Object.values(state).filter(Boolean).length) {
			return
		}

		console.log(Object.values(state).filter(Boolean).length)
		localStorage.setItem(localStorageKey, JSON.stringify(state))
	}, [state])

	return (
		<DistributionStateContext.Provider value={state}>
			<DistributionDispatchContext.Provider value={dispatch}>
				{children}
			</DistributionDispatchContext.Provider>
		</DistributionStateContext.Provider>
	)
}

function useDistributionState() {
	const context = useContext(DistributionStateContext)

	if (context === undefined) {
		throw new Error(
			'useDistributionState must be used within a DistributionProvider'
		)
	}
	return context
}

function useDistributionDispatch() {
	const context = useContext(DistributionDispatchContext)

	if (context === undefined) {
		throw new Error(
			'useDistributionDispatch must be used within a DistributionProvider'
		)
	}
	return context
}

export {
	DistributionProvider,
	useDistributionState,
	useDistributionDispatch,
	DistributionTypes,
}
