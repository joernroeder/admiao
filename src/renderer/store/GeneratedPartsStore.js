import React, { createContext, useContext, useReducer } from 'react'

const initialState = {
	parts: [],
	confirmed: [],
}

function reducer(state, action) {
	switch (action.type) {
		case 'SET_PARTS': {
			console.log('setting parts in store', action)
			return {
				...state,
				confirmed: Array(action.payload.length).fill(false),
				parts: action.payload,
			}
		}

		case 'CONFIRMED': {
			const updatedConfirmed = {
				confirmed: state.confirmed.map((isConfirmed, index) => {
					if (index === action.payload) {
						return true
					}

					return isConfirmed
				}),
			}

			return {
				...state,
				...updatedConfirmed,
			}
		}

		case 'RESET': {
			return {
				...initialState,
			}
		}

		default:
			return state
	}
}

const GeneratedPartsStateContext = createContext(undefined)
const GeneratedPartsDispatchContext = createContext(undefined)

function GeneratedPartsProvider({ children }) {
	const [state, dispatch] = useReducer(reducer, initialState)

	return (
		<GeneratedPartsStateContext.Provider value={state}>
			<GeneratedPartsDispatchContext.Provider value={dispatch}>
				{children}
			</GeneratedPartsDispatchContext.Provider>
		</GeneratedPartsStateContext.Provider>
	)
}

function useGeneratedParts() {
	const context = useContext(GeneratedPartsStateContext)

	if (context === undefined) {
		throw new Error(
			'useGeneratedParts must be used within a GeneratedPartsProvider'
		)
	}
	return context
}

function useGeneratedPartsDispatch() {
	const context = useContext(GeneratedPartsDispatchContext)

	if (context === undefined) {
		throw new Error(
			'useDistributionDispatch must be used within a GeneratedPartsProvider'
		)
	}
	return context
}

export { GeneratedPartsProvider, useGeneratedParts, useGeneratedPartsDispatch }
