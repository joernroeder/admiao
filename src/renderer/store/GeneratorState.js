import React, { createContext, useContext, useState } from 'react'

const STATES = {
	loading: 'LOADING',
	successfullyGenerated: 'SUCCESSFULLY_GENERATED',
	generationFailed: 'GENERATION_FAILED',
	successfullyDistributed: 'SUCCESSFULLY_DISTRIBUTED',
	distributionFailed: 'DISTRIBUTION_FAILED',
}

const GeneratorStateContext = createContext(undefined)

function GeneratorStateProvider({ children }) {
	const [generatorState, setGeneratorState] = useState({
		state: undefined,
		pendingIdentifier: null,
		error: null,
		// additional data returned by the ipc methods may get stored here.
		// todo be more precise/refactor once all distribution types are implemented
		meta: {},
	})

	return (
		<GeneratorStateContext.Provider
			value={[generatorState, setGeneratorState]}
		>
			{children}
		</GeneratorStateContext.Provider>
	)
}

function useGeneratorState() {
	const context = useContext(GeneratorStateContext)

	if (context === undefined) {
		throw new Error(
			'useGeneratedParts must be used within a GeneratedPartsProvider'
		)
	}
	return context
}

export { GeneratorStateProvider, useGeneratorState, STATES }
