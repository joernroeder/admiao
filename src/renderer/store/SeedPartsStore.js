import React, { createContext, useContext, useEffect, useReducer } from 'react'

const localStorageKey = 'SeedPartsState'

const initialState = {
	uniquePartsN: 3,
	requiredPartsT: 2,

	partsPerDisk: 1,
	requiredDisks: 3,

	recoveredFromPreviousSelection: false,
	isInitialState: true,
}

const calcDisks = (uniquePartsN, requiredPartsT) => {
	const partsPerDisk = requiredPartsT - 1
	const requiredDisks = Math.ceil(uniquePartsN / partsPerDisk)

	return {
		partsPerDisk,
		requiredDisks,
	}
}

function reducer(state, action) {
	const { uniquePartsN, requiredPartsT } = state

	const isModifiedState = {
		recoveredFromPreviousSelection: false,
		isInitialState: false,
	}

	switch (action.type) {
		case 'INCREMENT_REQUIRED': {
			const updatedRequiredPartsT = Math.min(
				requiredPartsT + 1,
				uniquePartsN
			)

			return {
				...state,
				requiredPartsT: updatedRequiredPartsT,
				...calcDisks(state.uniquePartsN, updatedRequiredPartsT),
				...isModifiedState,
			}
		}

		case 'DECREMENT_REQUIRED': {
			const updatedRequiredPartsT = Math.max(requiredPartsT - 1, 1)

			return {
				...state,
				requiredPartsT: updatedRequiredPartsT,
				...calcDisks(state.uniquePartsN, updatedRequiredPartsT),
				...isModifiedState,
			}
		}

		case 'INCREMENT_UNIQUE': {
			const updatedUniquePartsN = uniquePartsN + 1

			return {
				...state,
				uniquePartsN: updatedUniquePartsN,
				...calcDisks(updatedUniquePartsN, state.requiredPartsT),
				...isModifiedState,
			}
		}

		case 'DECREMENT_UNIQUE': {
			const updatedUniquePartsN = Math.max(uniquePartsN - 1, 1)
			// also decrement required parts if they would exceed otherwise
			const updatedRequiredPartsT = Math.min(
				updatedUniquePartsN,
				requiredPartsT
			)

			return {
				...state,
				uniquePartsN: updatedUniquePartsN,
				requiredPartsT: updatedRequiredPartsT,
				...calcDisks(updatedUniquePartsN, updatedRequiredPartsT),
				...isModifiedState,
			}
		}

		case 'CONFIRM': {
			return {
				...state,
				...isModifiedState,
			}
		}

		case 'RESET': {
			localStorage.removeItem(localStorageKey)

			return {
				...initialState,
			}
		}

		default:
			return state
	}
}

const SeedPartsStateContext = createContext(undefined)
const SeedPartsDispatchContext = createContext(undefined)

function SeedPartsProvider({ children }) {
	// uses a previous state from local storage if available or uses the initial one defined above.
	const [state, dispatch] = useReducer(
		reducer,
		initialState,
		initialState => {
			const previousState = localStorage.getItem(localStorageKey)

			return previousState
				? {
						...JSON.parse(previousState),
						recoveredFromPreviousSelection: true,
				  }
				: initialState
		}
	)

	// store current state in local storage.
	// This is used to improve the backup process if the user paused to print the
	// seed templates.
	useEffect(() => {
		if (state.isInitialState) {
			return
		}

		localStorage.setItem(localStorageKey, JSON.stringify(state))
	}, [state])

	return (
		<SeedPartsStateContext.Provider value={state}>
			<SeedPartsDispatchContext.Provider value={dispatch}>
				{children}
			</SeedPartsDispatchContext.Provider>
		</SeedPartsStateContext.Provider>
	)
}

function useSeedPartsState() {
	const context = useContext(SeedPartsStateContext)

	if (context === undefined) {
		throw new Error(
			'useSeedPartsState must be used within a SeedPartsProvider'
		)
	}
	return context
}

function useSeedPartsDispatch() {
	const context = useContext(SeedPartsDispatchContext)

	if (context === undefined) {
		throw new Error(
			'useSeedPartsDispatch must be used within a SeedPartsProvider'
		)
	}
	return context
}

export { SeedPartsProvider, useSeedPartsState, useSeedPartsDispatch }
