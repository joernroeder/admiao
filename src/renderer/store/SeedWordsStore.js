import React, { createContext, useContext, useReducer } from 'react'
const bip39 = require('bip39')

const initialWords = [
	/*
	'answer',
	'dice',
	'warm',
	'cup',
	'shock',
	'orbit',
	'pistol',
	'seat',
	'modify',
	'matrix',
	'genre',
	'renew',
	*/
]

const initialState = {
	words: initialWords,
	total: 12, // todo handle other seeds 18, 24
	isConfirmed: false,
	isValid: getIsValid(initialWords),
}

function getIsValid(words) {
	return bip39.validateMnemonic(words.join(' '))
}

function wordsReducer(state, action) {
	const { words } = state

	switch (action.type) {
		case 'APPEND': {
			const updatedWords = [...words, action.payload]

			return {
				...state,
				words: updatedWords,
				isValid: getIsValid(updatedWords),
			}
		}

		case 'RESET': {
			return {
				...state,
				words: [],
				isConfirmed: false,
				isValid: false,
			}
		}

		case 'CLEANUP_SEED': {
			return {
				...state,
				words: [],
			}
		}

		case 'POP': {
			const updatedWords = [...words.slice(0, -1)]

			return {
				...state,
				words: updatedWords,
				isValid: getIsValid(updatedWords),
			}
		}

		case 'CONFIRM': {
			return {
				...state,
				isConfirmed: true,
			}
		}

		default:
			return state
	}
}

const SeedWordsStateContext = createContext(undefined)
const SeedWordsDispatchContext = createContext(undefined)

function SeedWordsProvider({ children }) {
	const [state, dispatch] = useReducer(wordsReducer, initialState)

	return (
		<SeedWordsStateContext.Provider value={state}>
			<SeedWordsDispatchContext.Provider value={dispatch}>
				{children}
			</SeedWordsDispatchContext.Provider>
		</SeedWordsStateContext.Provider>
	)
}

function useSeedWordsState() {
	const context = useContext(SeedWordsStateContext)

	if (context === undefined) {
		throw new Error(
			'useSeedWordsState must be used within a SeedWordsProvider'
		)
	}
	return context
}

function useSeedWordsDispatch() {
	const context = useContext(SeedWordsDispatchContext)

	if (context === undefined) {
		throw new Error(
			'useSeedWordsDispatch must be used within a SeedWordsProvider'
		)
	}
	return context
}

export {
	SeedWordsProvider,
	useSeedWordsState,
	useSeedWordsDispatch,
	wordsReducer,
}
