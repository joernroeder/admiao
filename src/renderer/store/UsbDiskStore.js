import React, { createContext, useContext, useReducer } from 'react'

const initialState = {
	useManagedDisks: true,
}

function reducer(state, action) {
	switch (action.type) {
		case 'USE_PREPARED_DISKS': {
			return {
				...state,
				useManagedDisks: true,
			}
		}

		case 'DISABLE_DISK_MANAGEMENT': {
			return {
				...state,
				useManagedDisks: false,
			}
		}

		default:
			return state
	}
}

const UsbDiskStateContext = createContext(undefined)
const UsbDiskDispatchContext = createContext(undefined)

function UsbDiskProvider({ children }) {
	const [state, dispatch] = useReducer(reducer, initialState)

	return (
		<UsbDiskStateContext.Provider value={state}>
			<UsbDiskDispatchContext.Provider value={dispatch}>
				{children}
			</UsbDiskDispatchContext.Provider>
		</UsbDiskStateContext.Provider>
	)
}

function useUsbDiskState() {
	const context = useContext(UsbDiskStateContext)

	if (context === undefined) {
		throw new Error('useUsbDiskState must be used within a UsbDiskProvider')
	}
	return context
}

function useUsbDiskDispatch() {
	const context = useContext(UsbDiskDispatchContext)

	if (context === undefined) {
		throw new Error(
			'useUsbDiskDispatch must be used within a UsbDiskProvider'
		)
	}
	return context
}

export { UsbDiskProvider, useUsbDiskState, useUsbDiskDispatch }
