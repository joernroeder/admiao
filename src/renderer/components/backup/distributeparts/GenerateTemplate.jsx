import React, { useState, useEffect } from 'react'

import { useSeedPartsState } from '../../../store/SeedPartsStore'
import PrintTemplate from './PrintTemplate'
import CopyTemplateToUsb from './CopyTemplateToUsb'

const GenerateTemplate = ({
	defaultTemplateIdentifier = undefined,
	defaultPrintLocation = 'local',
}) => {
	const { requiredPartsT, uniquePartsN } = useSeedPartsState()
	const { ipcRenderer } = window
	const [templateIdentifier, setTemplateIdentifier] = useState(
		defaultTemplateIdentifier
	)
	const [templateError, setTemplateError] = useState()
	const [printLocation, setPrintLocation] = useState(defaultPrintLocation)

	/*
	const removeTemplate = () => {
		if (!templateIdentifier) {
			return
		}

		ipcRenderer.send('remove-template', templateIdentifier)
	}
	*/

	// generate pdf
	useEffect(() => {
		if (templateIdentifier) {
			return
		}

		// set the identifier once we got the "template-generated" message from the main process
		const onTemplateGenerated = (event, { error, data }) => {
			if (error) {
				return setTemplateError(error)
			}

			const { identifier } = data

			setTemplateIdentifier(identifier)
		}

		ipcRenderer.on('template-generated', onTemplateGenerated)

		ipcRenderer.send('generate-template', {
			requiredPartsT,
			uniquePartsN,
		})

		return () => {
			ipcRenderer.removeListener(
				'template-generated',
				onTemplateGenerated
			)
		}
	}, [ipcRenderer, uniquePartsN, requiredPartsT, templateIdentifier])

	// todo print template error
	if (templateError) {
		console.warn(templateError)
	}

	return (
		<>
			{printLocation === 'local' ? (
				<PrintTemplate
					templateIdentifier={templateIdentifier}
					onCopyToDrive={() => setPrintLocation('remote')}
				/>
			) : (
				<CopyTemplateToUsb templateIdentifier={templateIdentifier} />
			)}
		</>
	)
}

export default GenerateTemplate
