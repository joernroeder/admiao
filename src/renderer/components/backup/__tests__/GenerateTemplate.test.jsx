import React from 'react'

import { act, fireEvent, render } from '@testing-library/react'
import { SeedPartsProvider } from '../../../store/SeedPartsStore'

import GenerateTemplate from '../distributeparts/GenerateTemplate'

let onCopyToDrive = null
let testIdentifier = null

jest.mock('../distributeparts/PrintTemplate', () => {
	return ({ onCopyToDrive: callback, templateIdentifier }) => {
		onCopyToDrive = callback
		testIdentifier = templateIdentifier
		return 'PrintTemplateComponent'
	}
})

jest.mock('../distributeparts/CopyTemplateToUsb', () => {
	return () => 'CopyTemplateToUsbComponent'
})

describe('Print Templates', () => {
	beforeEach(() => {
		//jest.resetModules();
		//jest.resetAllMocks()
		onCopyToDrive = null
		testIdentifier = null

		global.ipcRenderer = {
			send: jest.fn(),
			once: jest.fn(),
			on: jest.fn(),
			removeListener: jest.fn(),
		}
	})

	const renderWithProviders = (children, renderFunction = render) => {
		return renderFunction(<SeedPartsProvider>{children}</SeedPartsProvider>)
	}

	test('it should render correctly the print template', () => {
		const { getByText } = renderWithProviders(
			<GenerateTemplate defaultPrintLocation={'local'} />
		)

		expect(getByText('PrintTemplateComponent')).toBeInTheDocument()
	})

	test('it should render correctly the copy to usb template', () => {
		const { getByText } = renderWithProviders(
			<GenerateTemplate defaultPrintLocation={'remote'} />
		)

		expect(getByText('CopyTemplateToUsbComponent')).toBeInTheDocument()
	})

	test('it should correctly update whenever the onCopyToDrive prop of the PrintTemplate component gets called', () => {
		const { getByText, container } = renderWithProviders(
			<GenerateTemplate />
		)

		expect(getByText('PrintTemplateComponent')).toBeInTheDocument()
		act(() => onCopyToDrive())
		expect(getByText('CopyTemplateToUsbComponent')).toBeInTheDocument()

		expect(container.firstChild).toMatchSnapshot()
	})

	test('it should send a `generate-template` message via ipc if no template identifier is set', () => {
		renderWithProviders(<GenerateTemplate />)

		expect(global.ipcRenderer.send).toBeCalledTimes(1)
		expect(global.ipcRenderer.send).toBeCalledWith('generate-template', {
			requiredPartsT: 2,
			uniquePartsN: 3,
		})

		renderWithProviders(
			<GenerateTemplate
				defaultTemplateIdentifier={'default-identifier'}
			/>
		)
		expect(global.ipcRenderer.send).toBeCalledTimes(1)
	})

	test('it should correctly update the template identifier once received via ipc', () => {
		renderWithProviders(<GenerateTemplate />)

		expect(global.ipcRenderer.on).toBeCalledTimes(1)
		const [channel, callback] = global.ipcRenderer.on.mock.calls[0]

		expect(channel).toEqual('template-generated')

		act(() => {
			callback(null, {
				error: null,
				data: {
					identifier: 'abc',
				},
			})
		})

		expect(testIdentifier).toEqual('abc')
	})
})
