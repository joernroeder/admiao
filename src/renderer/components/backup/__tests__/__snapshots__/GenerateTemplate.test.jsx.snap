// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Print Templates it should correctly update whenever the onCopyToDrive prop of the PrintTemplate component gets called 1`] = `CopyTemplateToUsbComponent`;
